posteriors/vi/dense.py

```diff
@@ -0,0 +1,287 @@
+from functools import partial
+import torch
+from torch.func import grad_and_value, vmap
+from optree import tree_map
+from optree.integration.torch import tree_ravel
+import torchopt
+
+from posteriors.types import TensorTree, Transform, LogProbFn
+from posteriors.tree_utils import tree_size, tree_insert_
+from posteriors.utils import (
+    is_scalar,
+    CatchAuxError,
+    L_from_flat,
+    L_to_flat,
+)
+
+
+def build(
+    log_posterior: Callable[[TensorTree, Any], float],
+    optimizer: torchopt.base.GradientTransformation,
+    temperature: float = 1.0,
+    n_samples: int = 1,
+    stl: bool = True,
+    init_cov: torch.Tensor | float = 1.0,
+) -> Transform:
+    """Builds a transform for variational inference with a Normal
+    distribution over parameters.
+
+    Find $\\mu$ and $\\Sigma$ that mimimize $\\text{KL}(N(θ| \\mu, \\Sigma) || p_T(θ))$
+    where $p_T(θ) \\propto \\exp( \\log p(θ) / T)$ with temperature $T$.
+
+    The log posterior and temperature are recommended to be [constructed in tandem](../../log_posteriors.md)
+    to ensure robust scaling for a large amount of data.
+
+    For more information on variational inference see [Blei et al, 2017](https://arxiv.org/abs/1601.00670).
+
+    Args:
+        log_posterior: Function that takes parameters and input batch and
+            returns the log posterior (which can be unnormalised).
+        optimizer: TorchOpt functional optimizer for updating the variational
+            parameters. Make sure to use lower case like torchopt.adam()
+        temperature: Temperature to rescale (divide) log_posterior.
+        n_samples: Number of samples to use for Monte Carlo estimate.
+        stl: Whether to use the stick-the-landing estimator
+            from (Roeder et al](https://arxiv.org/abs/1703.09194).
+        init_cov: Initial covariance matrix of the variational distribution.
+
+    Returns:
+        Dense VI transform instance.
+    """
+    init_fn = partial(init, optimizer=optimizer, init_cov=init_cov)
+    update_fn = partial(
+        update,
+        log_posterior=log_posterior,
+        optimizer=optimizer,
+        temperature=temperature,
+        n_samples=n_samples,
+        stl=stl,
+    )
+    return Transform(init_fn, update_fn)
+
+
+class VIDenseState(NamedTuple):
+    """State encoding a diagonal Normal variational distribution over parameters.
+
+    Attributes:
+        params: Mean of the variational distribution.
+        cov: Covariance matrix of the variational distribution.
+        L_factor: Flat representation of the nonzero values of the lower
+            triangular matrix $L$ satisfying $LL^T$ = cov.
+        opt_state: TorchOpt state storing optimizer data for updating the
+            variational parameters.
+        nelbo: Negative evidence lower bound (lower is better).
+        aux: Auxiliary information from the log_posterior call.
+    """
+
+    params: TensorTree
+    cov: torch.Tensor
```

Comment:
I think we can reduce memory by not including the cov in the state since its not really needed in update and simplifies things somewhat. One can always reconstruct cov from L_factor quite easily (note sample will have to be changed too).

Reply:
@SamDuffield Thank you for your response. 

I have removed the cov tracking and updated documentation, tests, update rules, and sampling as needed. I have also removed the need to specify the width of the lower triangular matrix reconstruction in the "L_from_flat" utils function.

Please let me know if anything else needs changing, and I'll get back to you as soon as I can. Thank you!